src/main/java/CRDT.java

```diff
@@ -1,21 +1,67 @@
-import java.io.Serializable;
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+import java.util.List;
 
-import com.ericsson.otp.erlang.OtpErlangBinary;
+import java.lang.reflect.Method;
 
-interface CRDT extends Serializable {
-    void invoke(GenericFunction obj);
+abstract class CRDT implements Antidote_interface {
+  private static final long serialVersionUID = 1L;
 
-    Object read();
+  abstract public Object value();
 
-    void snapshot();
+  public void invoke(GenericFunction obj) {
+    try {
+      String method_name = obj.getFunctionName();
+      List<Object> args = obj.getArguments();
+      /* Integer id = obj.getId(); */
 
-    // just serialize for snapshot read
+      Class<?>[] argTypes = (Class[]) this.getClass().getField(method_name).get(this);
 
+      Method method;
+      try {
+        method = this.getClass().getDeclaredMethod(method_name, argTypes);
+
+        method.invoke(this, args.toArray());
+      } catch (NoSuchMethodException e) {
+        // This is the case where the method has generic args
+        // Still doesn't work great
+        for (int i = 0; i < args.size(); i++) {
+          argTypes[i] = Object.class;
+        }
+
+        method = this.getClass().getDeclaredMethod(method_name, argTypes);
+        method.invoke(this, args.toArray());
```

Comment:
we can ask andru for help here, but I'm sure we can think of something "in-person" too

Reply:
Outdated as far as I can tell since we are now told the types of the method via fields.

Reply:
then remove away!